src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -0,0 +1,244 @@
+  ObjectCacheFile cache_file(m_cct, cache_dir + "/" + cache_file_name);
+
+  ret = cache_file.write_object_to_file(*read_buf, file_size);
+  if (ret < 0) {
+    lderr(m_cct) << "fail to write cache file" << dendl;
+
+    m_policy->update_status(cache_file_name, OBJ_CACHE_NONE);
+    delete read_buf;
+    return ret;
+  }
+
+  // update metadata
+  assert(OBJ_CACHE_SKIP == m_policy->get_status(cache_file_name));
+  m_policy->update_status(cache_file_name, OBJ_CACHE_PROMOTED);
+  assert(OBJ_CACHE_PROMOTED == m_policy->get_status(cache_file_name));
+
+  delete read_buf;
+  return ret;
+
+  evict_objects();
+}
+
+int ObjectCacheStore::lookup_object(std::string pool_name,
+                                    std::string object_name) {
+  ldout(m_cct, 20) << "object name = " << object_name
+                   << " in pool: " << pool_name << dendl;
+
+  int pret = -1;
+  cache_status_t ret = m_policy->lookup_object(pool_name + object_name);
+
+  switch(ret) {
+    case OBJ_CACHE_NONE: {
+      pret = do_promote(pool_name, object_name);
+      if (pret < 0) {
+        lderr(m_cct) << "fail to start promote" << dendl;
+      }
+      return -1;
+    }
+    case OBJ_CACHE_PROMOTED:
+      return 0;
+    case OBJ_CACHE_SKIP:
+      return -1;
+    default:
+      lderr(m_cct) << "unrecognized object cache status." << dendl;
+      assert(0);
+  }
+}
+
+int ObjectCacheStore::promote_object(librados::IoCtx* ioctx,
+                                     std::string object_name,
+                                     librados::bufferlist* read_buf,
+                                     Context* on_finish) {
+  ldout(m_cct, 20) << "object name = " << object_name << dendl;
+
+  auto ctx = new FunctionContext([on_finish](int ret) {
+    on_finish->complete(ret);
+  });
+
+  librados::AioCompletion* read_completion = create_rados_callback(ctx);
+  // issue a zero-sized read req to get full obj
+  int ret = ioctx->aio_read(object_name, read_completion, read_buf, 0, 0);
+  if(ret < 0) {
+    lderr(m_cct) << "failed to read from rados" << dendl;
+  }
+  read_completion->release();
+
+  return ret;
+}
+
+int ObjectCacheStore::evict_objects() {
+  ldout(m_cct, 20) << dendl;
+
+  std::list<std::string> obj_list;
+  m_policy->get_evict_list(&obj_list);
+  for (auto& obj: obj_list) {
+    do_evict(obj);
+  }
+}
+
+int ObjectCacheStore::do_evict(std::string cache_file) {
+  ldout(m_cct, 20) << "file = " << cache_file << dendl;
+
+  //TODO(): need a better way to get file path
+
+  std::string cache_dir = m_cache_root_dir;
+
+   if (m_dir_num > 0) {
+    auto const pos = cache_file.find_last_of('.');
+    cache_dir = cache_dir + "/" + std::to_string(stoul(cache_file.substr(pos+1)) % m_dir_num);
+  }
+  std::string cache_file_path = cache_dir + "/" + cache_file;
+
+  ldout(m_cct, 20) << "delete file: " << cache_file_path << dendl;
+  int ret = std::remove(cache_file_path.c_str());
+   // evict entry in policy
+  if (ret == 0) {
+    m_policy->evict_entry(cache_file);
+  }
```

Comment:
It looks like a  race is possible here if in the window between object is removed and the entry is evicted a new request comes and will try to read from the deleted file. Probably, before removing the file change the entry status to `OBJ_CACHE_SKIP`?